### ISSUE TITLE
Loss metrics do not have gradient calculations?..

### ISSUE THREAD

**Issue by Strateus** [apache/incubator-mxnet#4312]
Im trying to figure out how gradient in the loss layer being calculated in mxnet, and still failing, documentation does not help. Can you please describe this logic or give me some link?
What i wanted to do is to define custom numpy loss layer, but its intefrace does not allow me to give out gradients wrt inputs, meaning i wont be able to control this when i calculate my loss. Or i miss some part in how mxnet is implemented? I've read about splitting SoftmaxLoss to SoftmaxOutput and something else, but that "something else" seems to be missing in metrics.py. For now i switched to using Accuracy metric, but it does not ideally fit my needs, so i wanted to modify it a bit, but modifying it without doing gradient calcs kind of does not make sense? Could you please guide me?

**Comment by Strateus** [comment]
Anyone can help?

**Comment by marugari** [comment]
R custom loss function tutorial
https://github.com/dmlc/mxnet/blob/master/docs/tutorials/r/CustomLossFunction.md

However, I think the accuracy is unsuitable because it is not differentiable.

**Comment by Strateus** [comment]
I dont see how i calculate and return gradient in this example? Can you please clarify? We need to have gradient at Loss layer, otherwise training does not make sense.

**Comment by marugari** [comment]
MXNet knows the derivative function of the `mx.sym.square` and elementwize minus.
The gradient of loss function can be got for Automatic Differentiation.

**Comment by Strateus** [comment]
Well, what if, for example, I have 1st item in the batch dependent in terms of gradient on 2nd item in a batch? Or if I for example have multi class problem and want to return gradient for labeled class and neighbor classes? How can mxnet automatically calculate that gradient if I do my calculations inside loss function? That does not make sense.

**Comment by Strateus** [comment]
Can please developers clarify?

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
